src/validation.cpp

```diff
@@ -350,21 +350,29 @@ void CChainState::MaybeUpdateMempoolForReorg(
     // the disconnectpool that were added back and cleans up the mempool state.
     m_mempool->UpdateTransactionsFromBlock(vHashUpdate);
 
-    const auto check_final_and_mature = [this, flags=STANDARD_LOCKTIME_VERIFY_FLAGS](CTxMemPool::txiter it)
+    // Check whether the transaction is still final and, if it spends a coinbase output, mature.
+    // Update the entry's cached LockPoints if needed. If this returns true, the tx would be invalid
+    // in the next block, so the mempool must remove this entry and all of its descendants.
+    const auto filter_final_and_mature = [this, flags=STANDARD_LOCKTIME_VERIFY_FLAGS](CTxMemPool::txiter it)
         EXCLUSIVE_LOCKS_REQUIRED(m_mempool->cs, ::cs_main) {
-        bool should_remove = false;
         AssertLockHeld(m_mempool->cs);
         AssertLockHeld(::cs_main);
         const CTransaction& tx = it->GetTx();
         LockPoints lp = it->GetLockPoints();
         const bool validLP{TestLockPointValidity(m_chain, lp)};
         CCoinsViewMemPool view_mempool(&CoinsTip(), *m_mempool);
-        if (!CheckFinalTx(m_chain.Tip(), tx, flags)
-            || !CheckSequenceLocks(m_chain.Tip(), view_mempool, tx, flags, &lp, validLP)) {
-            // Note if CheckSequenceLocks fails the LockPoints may still be invalid
-            // So it's critical that we remove the tx and not depend on the LockPoints.
-            should_remove = true;
-        } else if (it->GetSpendsCoinbase()) {
+        // The transaction must be final.
+        if (!CheckFinalTx(m_chain.Tip(), tx, flags)) return true;
+        // Note if CheckSequenceLocks fails the LockPoints may still be invalid
+        // So it's critical that we remove the tx and not depend on the LockPoints.
+        if (!CheckSequenceLocks(m_chain.Tip(), view_mempool, tx, flags, &lp, validLP)) {
+            return true;
+        } else if (!validLP) {
+            // CheckSequenceLocks updates lp. Update the mempool entry LockPoints.
+            m_mempool->mapTx.modify(it, update_lock_points(lp));
+        }
+        // If the transaction spends any coinbase outputs, check maturity.
+        if (it->GetSpendsCoinbase()) {
```

Comment:
Previously, this check and the body of the `if` would have been executed before updating the lock points. Now the order is reversed. Could the updated lock points change the outcome of any of this code, e.g. the `mapTx.find()` call?